### PR TITLE
brew.rb: tell users to fix head issues with inreplace

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
WIP to ensure not only build command failures tell users of `--HEAD` to make a pull request

### CODE REVIEW

**Review by carlocab — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Good idea 👍🏻

**Comment by Bo98** [comment]
Not convinced on making a parent `FormulaError` considering inreplace is not formula specific (it's used in bump-cask-pr etc. too). 

Probably best to keep the error as it was and either:
* just add a generic `object` variable, or;
* have Formula-specific `inreplace` and `inreplace_pairs` methods which call super but catches the error and converts it to a `BuildError` or something similar. This would allow a lot of `BuildError`-specific logic to be reused.

**Comment by carlocab** [comment]
Yea, catching the `inreplace` error throwing a `BuildError` seems much nicer.

**Comment by SMillerDev** [comment]
And that's why it was a draft, thanks @Bo98. I'll check that later

**Comment by SMillerDev** [comment]
A lot later, but it's done now.

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -2169,6 +2166,12 @@ def test_fixtures(file)
   # end</pre>
   def install; end
 
+  def inreplace(paths, before = nil, after = nil, audit_result = true) # rubocop:disable Style/OptionalBooleanParameter
```
Comment (carlocab) [inline Library/Homebrew/formula.rb:2169]: Can we keep the docstring?

**Review by carlocab — APPROVED** [review]

**Review by Bo98 — APPROVED** [review]
Fantastic. This is exactly how I hoped it would look.

**Review by MikeMcQuaid — APPROVED** [review]
Tiny style suggestion but: looks good to me!

**Comment by bayandin** [comment]
It looks like something went wrong, see https://github.com/Homebrew/homebrew-core/pull/107575:

There's no error in the install block, but a problem with `inreplace` appears in ` ==> ENV`  block

```
 ==> ENV
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build_environment.rb:60:in `keys': undefined method `keys' for nil:NilClass (NoMethodError)
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build_environment.rb:65:in `dump'
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/exceptions.rb:505:in `dump'
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:148:in `rescue in <main>'
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:136:in `<main>'
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2222:in `rescue in inreplace': Failed executing: inreplace configure (BuildError)
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2219:in `inreplace'
  	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/nvc.rb:44:in `install'
```

https://github.com/Homebrew/homebrew-core/runs/7729673651?check_suite_focus=true#step:7:93

**Pull request merged**